rules/firestore.json

```diff
@@ -35,50 +36,188 @@ service cloud.firestore {
+                             partHasNotStarted(resource.data);
+            // A number of fields are readonly:
+            // - typeGame
+            // - playerZero/playerOne once the part has started
+            // - beginning once the part has started
+            // the part is considered to be started as soon as beginning is set
+            // For any other fields, both players can modify them, with some restrictions:
+            // - lastUpdate must be updated monotonously by the corresponding player
+            // - turn can only be made +1 (usual turn), -1 (take back), or -2 (take back)
+            allow update: if userIsVerified() &&
+                             preservesField('typeGame') &&
+                             (preservesField('playerZero') || partHasNotStarted(resource.data)) &&
+                             (preservesField('playerOne') || partHasNotStarted(resource.data)) &&
+                             (preservesField('beginning') || partHasNotStarted(resource.data)) &&
+                             (preservesField('turn') || safeTurnUpdate(resource.data.turn, request.resource.data.turn)) &&
+                             (safeLastUpdate(resource.data.lastUpdate, request.resource.data.lastUpdate)) &&
+                             (userIsPlayer(resource.data) || (partHasNotStarted(resource.data) && userIsChosenOpponentInJoiner(id)));
+
+            // All verified user can read parts (needed for the lobby)
+            allow read: if userIsVerified();
+        }
+        function userIsChosenOpponentInJoiner(partId) {
+            let joiner = get(/databases/$(database)/documents/joiners/$(partId)).data;
+            return hasField(joiner, 'chosenOpponent') && joiner.chosenOpponent.id == request.auth.uid;
+        }
+        function userIsPlayer(part) {
+            let userIsPlayerZero = part.playerZero.id == request.auth.uid;
+            let userIsPlayerOne = hasField(part, 'playerOne') && part.playerOne.id == request.auth.uid;
+            return userIsPlayerZero || userIsPlayerOne;
+        }
+        function safeTurnUpdate(oldTurn, newTurn) {
+            return (newTurn == oldTurn + 1) || (newTurn == oldTurn - 1) || (newTurn == oldTurn - 2);
+        }
+        function safeLastUpdate(oldLastUpdate, newLastUpdate) {
+            let legitimatePlayerZeroUpdate = newLastUpdate.player == 0 && resource.data.playerZero.id == request.auth.uid;
+            let legitimatePlayerOneUpdate = hasField(resource.data, 'playerOne') && newLastUpdate.player == 1 && resource.data.playerOne.id == request.auth.uid;
+            let indexIncremented = newLastUpdate.index == oldLastUpdate.index + 1;
+            return (legitimatePlayerZeroUpdate || legitimatePlayerOneUpdate || partHasNotStarted(resource.data)) && indexIncremented;
+        }
+        function partHasNotStarted(part) {
+            return hasField(part, 'beginning') == false;
+        }
+        function hasTimedOut(creator) {
```

Comment:
rename to "creatorTimedOut" pour éviter d'être confusant, car par défaut on part plutôt du principe ici que le sujet c'est l'utilisateur envoyant cette requête que le créateur.

Reply:
ou alors hasTimedOut(user)